### DISCUSSION TITLE
make integration with nx.dev monorepo and angular cli builders possible

### DISCUSSION THREAD

**Discussion by fxck** [strapi/strapi#6221] [Legacy Topics]
// edit: created proof of concept PR - #6233

I'm trying to integrate strapi into nx monorepo. In theory it should be possible, in reality, not so much. There are two main problems:

1) strapi cli expects package.json to be on the same level as your strapi app
2) there's no way to define output destination for production build

With nx monorepo, your project structure looks something like this

![image](https://user-images.githubusercontent.com/1303561/82044164-1ee42c00-96ad-11ea-8fbe-ab4a57bf2648.png)

and you want your build to end up in `/dist`.

If only you could pass optional `--dir`, and `--outputPath` options to `strapi develop` / `strapi build / start` and take those in consideration, instead of hardcoding those paths, it should all work fine.

Thoughts?

**Comment by derrickmehaffy** [comment]
Hello,

I see you are wanting to ask a question that is not really a bug report or a feature request, questions should be directed to the [discussions](https://github.com/strapi/strapi/discussions)

I have converted this to a discussion for you, we try to keep Github clean and only for bug reports and feature requests.

**Reply by fxck** [reply]
Alright, thanks, I guess you might as well call it a discussion.

Now I'd be more than happy to try and make a PR, but the project is pretty big and I don't know what the all the implications it could have.

So far from what I can tell:

- ~~`/packages/strapi/bin/strapi` logic of checking whether you are in strapi project would have to change (since the project could be nested with package.json in root folder)~~

- `strapi develop` would have to accept `dir` as param with `process.cwd()` as the default

- `core/app-configuration` would need to additionally accept `root` option, with `dir` as the default

- `strapi build` would need to accept `outputPath` option, which could copy the whole folder after success to desired path (I don't think you can generate static build files, so you needs src along the build?)

- `strapi start` would have to accept `dir` option

**Reply by fxck** [reply]
actually scratch that first point, you would have `strapi` as `dependency` in your package.json anyway

**Comment by fxck** [comment]
https://github.com/strapi/strapi/pull/6233

**Comment by robertpiosik** [comment]
I'm not sure you are going in the right direction. Packages managed within monorepo should have scoped dependencies and it is easily achievable using yarn workspaces.

**Reply by fxck** [reply]
Inside what monorepo? Nx? I'm pretty sure you have shared dependencies there (there is an open issue https://github.com/nrwl/nx/issues/1777, but even if it were an option, I don't think it's always necessary - "We believe in the single-version policy, where, for instance, all applications using NestJS, use the same version of NestJS. Even though it looks a bit crazy at first, we think it is beneficial. Companies like Google do it. So having one node_modules at the root is very similar to what Google does"). 

All I basically want is for Strapi to be able to live in a nested directory (#6233).

**Reply by robertpiosik** [reply]
Hmm, I get yout idea but doesn't it lead to a giant `package.json` which is really hard to maintain? What if two or more of your packages uses different versions of some deps? I haven't heard about nx thing though.

**Reply by robertpiosik** [reply]
> "We believe in the single-version policy, where, for instance, all applications using NestJS, use the same version of NestJS. Even though it looks a bit crazy at first, we think it is beneficial. Companies like Google do it. So having one node_modules at the root is very similar to what Google does").

This is so abstract to me. I can't imagine how time consuming would be updating some minor thing in a little older package 😨

**Reply by fxck** [reply]
It's not time consuming at all. You try and stay away from such situations (I haven't yet encountered it once with 4 projects and 3 libraries). We have a lot of shared code between projects, makes sense to run the same version of everything everywhere. Given that Strapi basically has only one dependency and that is node, it's not much of a problem anyway.

**Reply by robertpiosik** [reply]
Is what you have proposed available in any other CLI tool? I would say it's a solid edge case. While utilising yarn workspace you could use a command `yarn workspace [package-name] [script-name]` to easily run any cli command in a given package scope.

In my opinion it's the cleanest approach to the problem.

**Comment by lauriejim** [comment]
Hello! I ping @Aurelsicoko and @alexandrebodin about this topic.

**Reply by fxck** [reply]
hello

**Comment by valeriocomo** [comment]
Hi everyone. I think monorepo-compliancy is a good improvement for Strapi. I'm interested in this topic and I suggest it could be very useful have the possibility to choose to be monorepo compliant or not

**Reply by asciidiego** [reply]
+1

**Comment by alikazai** [comment]
Hi i am also planning to do my new project this way as well as integrate strapi into an existing nx monorepo. any update if this ticket has been approved or when we can help out to implement such an integration. 

thanks

**Reply by fxck** [reply]
https://github.com/strapi/strapi/pull/6233#issuecomment-658736946

**Comment by adrian-ub** [comment]
any ideas?
